packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +240,135 @@ def setup_opt_parser
+      soa_record = query_dns.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+      existing_serial = soa_record.serial if soa_record
+      new_serial = (existing_serial + 1).to_s if existing_serial
+      {
+        existing_ip: existing_ip,
+        new_serial: new_serial
+      }
+    end
+
+    def parse_zone_file(zone_file)
+      run_cmd('rndc freeze')
+      zone_file_data = File.read(zone_file).split("\n")
+      # find the character A enclosed by any whitespace chars
+      a_record = zone_file_data.find { |line| !(line =~ /\sA\s/).nil? }
+      existing_ip = a_record.split('A').last.strip if a_record
+      serial_line = zone_file_data.find { |line| line.downcase.include?('serial') }
+      existing_serial = serial_line.split('Serial').first.to_i if serial_line
+      new_serial = (existing_serial + 1).to_s if existing_serial
+      {
+        existing_ip: existing_ip,
+        new_serial: new_serial
+      }
+    end
+
+    def get_dns_data
+      STDOUT.puts 'gathering DNS data...'
+      local_ip = '127.0.0.1'
+      zone = @scenario_answers['foreman_proxy']['dns_zone']
+      reverse_zone = @scenario_answers['foreman_proxy']['dns_reverse']
+      if reverse_zone.is_a?(Array)
+        reverse_zone = reverse_zone.first
+      end
+      old_fqdn = @old_hostname
+      soa_admin_domain = "root.#{zone}"
+      new_fqdn = @new_hostname
+      key_file = @scenario_answers['foreman_proxy']['keyfile']
+      begin
+        STDOUT.puts 'querying local DNS server...'
+        ip_serial_data = query_dns_server(old_fqdn, zone, local_ip)
+        existing_ip = ip_serial_data[:existing_ip]
+        new_serial = ip_serial_data[:new_serial]
+        new_reverse_serial = query_dns_server(old_fqdn, reverse_zone, local_ip)[:new_serial]
+      rescue Resolv::ResolvError => e
+        STDOUT.puts e
+        STDOUT.puts 'Parsing zone file as fallback'
+        ip_serial_data = parse_zone_file("/var/named/dynamic/db.#{zone}")
+        existing_ip = ip_serial_data[:existing_ip]
+        new_serial = ip_serial_data[:new_serial]
+        new_reverse_serial = parse_zone_file("/var/named/dynamic/db.#{reverse_zone}")[:new_serial]
+      ensure
+        run_cmd('rndc thaw')
+      end
+
+
+      {
+        local_ip: local_ip,
+        zone: zone,
+        old_fqdn: old_fqdn,
+        soa_admin_domain: soa_admin_domain,
+        existing_ip: existing_ip,
+        new_serial: new_serial,
+        new_reverse_serial: new_reverse_serial,
+        new_fqdn: new_fqdn,
+        key_file: key_file,
+        reverse_zone: reverse_zone
+      }
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_dns_records(dns_data = {})
+      dns_data.each do |key, value|
+        unless value
+          raise "Error gathering DNS data: couldn't find value for '#{key}'"
+        end
+      end
+
+      local_ip = dns_data[:local_ip]
```

Comment:
I'd like to see all these variable reassignments go away. You can do that by making `forward_dns_command` a method that takes required keyword arguments with a pattern like this:

```
hsh = {                                                                                                                                                       
  foo: 'its foo!',                                                                                                                                            
  bar: 'its bar!',                                                                                                                                            
}                                                                                                                                                             
                                                                                                                                                              
def expand_hash(foo:, bar:)                                                                                                                                   
  puts foo                                                                                                                                                    
  puts bar                                                                                                                                                    
end                                                                                                                                                           
                                                                                                                                                              
expand_hash(**hsh) 
```

outputs:

```
its foo!
its bar!
```

Reply:
:open_mouth:  This is what I was looking for!

Reply:
@jturel Looks like keyword arguments with default values were introduced in Ruby 2.0, but required keyword args like you outlined didn't [arrive](https://thoughtbot.com/blog/ruby-2-keyword-arguments) until Ruby 2.1.  :sob: